py/picca/delta_extraction/data_catalogues/desi_healpix.py

```diff
@@ -346,8 +207,181 @@ def read_file(self, filename, catalogue, forests_by_targetid):
+                Forest.mask_fields += ["reso"]
+            if "reso_pix" not in Forest.mask_fields:
+                Forest.mask_fields += ["reso_pix"]
+            if "resolution_matrix" not in Forest.mask_fields:
+                Forest.mask_fields += ["resolution_matrix"]
+
+    def __parse_config(self, config):
+        """Parse the configuration options
+
+        Arguments
+        ---------
+        config: configparser.SectionProxy
+        Parsed options to initialize class
+
+        Raise
+        -----
+        DataError upon missing required variables
+        """
+        self.use_non_coadded_spectra = config.getboolean(
+            "use non-coadded spectra")
+        if self.use_non_coadded_spectra is None:
+            raise DataError(
+                "Missing argument 'use non-coadded spectra' required by DesiHealpix"
+            )
+        self.num_processors = config.getint("num processors")
+        if self.num_processors is None:
+            raise DataError(
+                "Missing argument 'num processors' required by DesiHealpix")
+        if self.num_processors == 0:
+            self.num_processors = (multiprocessing.cpu_count() // 2)
+
+    def read_data(self):
+        """Read the data.
+
+        Method used to read healpix-based survey data.
+
+        Return
+        ------
+        is_mock: bool
+        False as DESI data are not mocks
+
+        is_sv: bool
+        True if all the read data belong to SV. False otherwise
+
+        Raise
+        -----
+        DataError if no quasars were found
+        """
+        in_nside = 64
+
+        healpix = [
+            healpy.ang2pix(in_nside,
+                           np.pi / 2 - row["DEC"],
+                           row["RA"],
+                           nest=True)
+            for row in self.catalogue
+        ]
+        self.catalogue["HEALPIX"] = healpix
+        self.catalogue.sort("HEALPIX")
+        grouped_catalogue = self.catalogue.group_by(["HEALPIX", "SURVEY"])
+
+        is_sv = True
+        forests_by_targetid = {}
+
+        arguments = []
+        for group in grouped_catalogue.groups:
+            healpix, survey = group["HEALPIX", "SURVEY"][0]
+
+            if survey not in ["sv", "sv1", "sv2", "sv3"]:
+                is_sv = False
+
+            #TODO: not sure if we want the dark survey to be hard coded in here, probably won't run on anything else, but still
+            input_directory = f'{self.input_directory}/{survey}/dark'
+            coadd_name = "spectra" if self.use_non_coadded_spectra else "coadd"
+            filename = (
+                f"{input_directory}/{healpix//100}/{healpix}/{coadd_name}-{survey}-"
+                f"dark-{healpix}.fits")
+
+            arguments.append((filename,group))
+
+        self.logger.info(f"reading data from {len(arguments)} files")
+
+        if self.num_processors>1:
+            with multiprocessing.Pool(processes=self.num_processors) as pool:
```

Comment:
I think the only part that really needs to be seperated for single/multi-core processing is the `imap_it=` vs the `reader=` line, the merging could be outside the if clause (minimizing code duplication)

Reply:
i.e. maybe just do `imap_it=(reader(arg) for arg in arguments)` in the sequential version and have the merging for loop outside

Reply:
We can also not check for number of processors. It should behave like a serial code when `num_processors=1`. I don't know if that would create some weird behavior

Reply:
the issue here is that it still would be run through multiprocessing stuff which makes debugging painful even if only on a single CPU, so having the serial case multiprocessing-free is just a convenience to allow debugging issues more easily

Reply:
Ok that makes sense. I'm worried about reducing the code duplication though. You can't take imap outside of `with`. For sequential version, keeping all in a list as you suggest might be memory intense. Let's postpone this to later testing if that's ok?